dd-sdk-android-core/api/apiSurface

```diff
@@ -323,14 +323,14 @@ interface com.datadog.android.core.persistence.Serializer<T: Any>
 fun <T: Any> Serializer<T>.serializeToByteArray(T, com.datadog.android.api.InternalLogger): ByteArray?
 data class com.datadog.android.core.persistence.datastore.DataStoreContent<T: Any>
   constructor(Int, T?)
-class com.datadog.android.core.sampling.RateBasedSampler : Sampler
+open class com.datadog.android.core.sampling.RateBasedSampler<T: Any> : Sampler<T>
   constructor(() -> Float)
   constructor(Float)
   constructor(Double)
-  override fun sample(): Boolean
+  override fun sample(T): Boolean
   override fun getSampleRate(): Float
-interface com.datadog.android.core.sampling.Sampler
-  fun sample(): Boolean
+interface com.datadog.android.core.sampling.Sampler<T: Any>
+  fun sample(T): Boolean
```

Comment:
this is potentially a binary-incompatible change, because public constructors of `DatadogInterceptor` and `TracingInterceptor` (and builder methods since recently) take `Sampler` instance which can be some customer-defined `Sampler` rather than default `RateBasedSampler`.

Is there a way to avoid breaking it?

Reply:
I hesitated to do it, as indeed it's technically a breaking change on a public interface. That being said, looking at our code base, the interface is used in: 
- the `internal class TelemetryEventHandler` constructor, but the sampler is built internally based on a sample rate provided in the configuration
- the `internal class SessionReplayFeature` constructor, but the sampler is built internally based on a sample rate provided in the configuration
- the `internal class DatadogLogHandler` constructor, but the sampler is built internally based on a sample rate provided in the configuration
- an internal property of the `internal class WebViewLogEventConsumer`, built internally based on a sample rate provided in the configuration
- in the `DatadogInterceptor` and `TracingInterceptor` **deprecated** constructors; they were replaced by a Builder. The builder only accepts a Sampler, which is the only place we could have ab issue for customer, which I guess is acceptable. I'll also add a builder method to set the sampling rate as I feel it's missing.

Reply:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**🎯 suggestion:** ‏ What about a keeping the builder method that accepts the `Sampler` and wrapping this old `Sampler` that the customer is passing to your own `Sampler<Unit>` ? In this case it will not be a breaking change anymore. And we could deprecate that method for later.

Reply:
will it make sense to make the API be like `fun sample(T? = null): Boolean` instead of `fun sample(T): Boolean`?

Reply:
I believe we probably don't want to allow nullable types in the interface definition. But probably even if we do so, it would be still binary incompatible change, because then customer implementing custom `Sampler` still need to update the method signature.

My main concern is indeed in the interceptors usage (other places are pretty much internal), but if we ultimately cannot avoid this and we need to progress on this topic, I'm fine with this minor breaking change.